[Feature] implement multi-warehouse

crossoverJie opened the pull request:
## Why I'm doing:

## What I'm doing:

- Add support for creating, dropping, and managing multiple warehouses
- Implement MultipleWarehouse class to handle worker group associations
- Add replay and save/load functionality for warehouses
- Include unit tests for multi-warehouse functionality


## What type of PR is this:

- [ ] BugFix
- [x] Feature
- [ ] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

Does this PR entail a change in behavior?

- [ ] Yes, this PR will result in a change in behavior.
- [x] No, this PR will not result in a change in behavior.

If yes, please specify the type of change:

- [ ] Interface/UI changes: syntax, type conversion, expression evaluation, display information
- [ ] Parameter changes: default values, similar parameters but with different default values
- [ ] Policy changes: use new policy to replace old one, functionality automatically enabled
- [ ] Feature removed
- [ ] Miscellaneous: upgrade & downgrade compatibility, etc.

## Checklist:

- [x] I have added test cases for my bug fix or my new feature
- [x] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function
- [ ] This is a backport pr

## Bugfix cherry-pick branch check:
- [ ] I have checked the version labels which the pr will be auto-backported to the target branch
  - [ ] 4.1
  - [ ] 3.5
  - [ ] 3.4
  - [ ] 3.3

crossoverJie committed:
feat(warehouse): implement multi-warehouse functionality

- Add support for creating, dropping, and managing multiple warehouses
- Implement MultipleWarehouse class to handle worker group associations
- Update WarehouseManager to support multi-warehouse operations
- Add replay and save/load functionality for warehouses
- Include unit tests for multi-warehouse functionality

Signed-off-by: crossoverJie <crossoverJie@gmail.com>

crossoverJie committed:
test(warehouse): add unit test for warehouse image save and load

- Implement clearWarehouse method in WarehouseManager for testing purposes
- Add testWarehouseImage test case in MultiWarehouseTest
- Verify warehouse data persistence across save and load operations

Signed-off-by: crossoverJie <crossoverJie@gmail.com>

crossoverJie committed:
Merge branch 'main' into support-multi-warehouse

crossoverJie committed:
fix(warehouse): clean up worker group when deleting multiple warehouse

- Add cleanup logic for associated worker groups when deleting a warehouse
- Implement getWorkerGroupId method in MultipleWarehouse class
- Update comments and improve code readability

Signed-off-by: crossoverJie <crossoverJie@gmail.com>

crossoverJie committed:
test(fe-core): update product feature test assertion

- Change the expected number of features from 1 to 2

Signed-off-by: crossoverJie <crossoverJie@gmail.com>

helifu commented on fe/fe-core/src/main/java/com/starrocks/feature/ProductFeature.java:
```diff
@@ -57,6 +57,11 @@ public String getLink() {
         ));
+        features.add(new ProductFeature(
+                "multi-warehouse",
+                "support multiple warehouses, each with its own worker group and configuration.",
```
Thanks for this PR 👍 
I have a questions about `its own worker group and configuration`: how can we inject the computer nodes for this warehouse?

crossoverJie commented on fe/fe-core/src/main/java/com/starrocks/feature/ProductFeature.java:
```diff
@@ -57,6 +57,11 @@ public String getLink() {
         ));
+        features.add(new ProductFeature(
+                "multi-warehouse",
+                "support multiple warehouses, each with its own worker group and configuration.",
```
This feature requires the use of an operator. After the [CRD](https://github.com/StarRocks/starrocks-kubernetes-operator/blob/main/doc/deploy_warehouse_howto.md) is created, the operator will automatically create CN nodes.

https://github.com/StarRocks/starrocks/blob/afbb3ca617e304f5ff78b84a4e14316eab1774f7/docker/dockerfiles/be/cn_entrypoint.sh#L83-L87

> each with its own worker group and configuration

This was generated by AI; specific edits will be made when it is officially merged.

crossoverJie committed:
Merge branch 'main' into support-multi-warehouse

crossoverJie committed:
feat(warehouse): add fallback to default warehouse when specified warehouse not found

- Introduced new configuration option `enable_rollback_default_warehouse`
- Implemented fallback logic in `WarehouseManager` for null warehouse scenarios
- Updated `getWarehouse` and `getWarehouseAllowNull` methods to support fallback

Signed-off-by: crossoverJie <crossoverJie@gmail.com>

crossoverJie committed:
Merge branch 'main' into support-multi-warehouse

crossoverJie committed:
fix(compute): validate warehouse existence before dropping compute node

Signed-off-by: crossoverJie <crossoverJie@gmail.com>

LiShuMing commented:
Thanks for your contributions.

We have already had another implemention for multi wareshouse support and will be soon released in the short future.

It will contain many other complete/product-ready functions.  Can you merge this with it later?

Naraujo13 commented:
Hi @LiShuMing !

We were looking for multi-warehouse - an essential feature for us - and I saw it is in the 2026 Roadmap and found this MR mentioning a short future release 9 months ago. 

Do you've any update or information to share on this? 
A link to the other MR mentioned or a rough update on how the plan is going would be very helpful so we can better understand it.

Thanks!
